dmutils/email/dm_notify.py

```diff
@@ -0,0 +1,123 @@
+    ):
+        """Set up logging and mail client."""
+        self.logger = logger or current_app.logger
+        self.client = self._client_class(govuk_notify_api_key, govuk_notify_base_url)
+
+    def get_all_notifications(self, **kwargs):
+        """Wrapper for notifications_python_client.notifications.NotificationsAPIClient::get_all_notifications"""
+        return self.client.get_all_notifications(**kwargs)['notifications']
+
+    def get_delivered_notifications(self, **kwargs):
+        """Wrapper for notifications_python_client.notifications.NotificationsAPIClient::get_all_notifications"""
+        return self.get_all_notifications(status='delivered', **kwargs)
+
+    def get_delivered_references(self, invalidate_cache=False):
+        """Get the references of all notifications that have already been delivered."""
+        if invalidate_cache:
+            self._sent_references_cache = None
+        if self._sent_references_cache is None:
+            self._sent_references_cache = [i['reference'] for i in self.get_delivered_notifications()]
+        return self._sent_references_cache
+
+    def _update_cache(self, reference):
+        """If the cache has been instantiated then cache the new reference."""
+        if self._sent_references_cache is not None:
+            self._sent_references_cache.append(reference)
+
+    def has_been_sent(self, reference):
+        """Checks for a matching reference in our list of delivered references."""
+        return reference in self.get_delivered_references()
+
+    @staticmethod
+    def get_reference(email_address, template_id, personalisation=None, **kwargs):
+        """
+        Method to return the standard reference given the variables the email is sent with.
+
+        :param email_address: Emails recipient
+        :param template_id: Emails template ID on Notify
+        :param personalisation: Template parameters
+        :param kwargs: Extra data passed to the reference eg. {'notes': 'Manual resend'}
+        :return: Hashed string 'reference' to be passed to client.send_email_notification or self.send_email
+        """
+        details_dict = OrderedDict([
+            ('email_address', email_address),
+            ('template_id', template_id),
+            ('personalisation', ','.join(stringify_dict(personalisation))),
+        ])
+        details_string = '|'.join(stringify_dict(details_dict))
+
+        return hash_string(details_string)
+
+    @staticmethod
+    def get_error_message(email_address, error):
+        """Format a logical error message from the error response."""
+        messages = []
+        message_prefix = 'Error sending message to {email_address}: '.format(email_address=email_address)
+        message_string = '{status_code} {error}: {message}'
+
+        for message in error.message:
+            format_kwargs = {
+                'status_code': error.status_code,
+                'error': message['error'],
+                'message': message['message'],
```

Comment:
So we either only log status and the error or we do a replace on the message with `message['message'].replace(email_address, '***')`